### DISCUSSION TITLE
Plugins sorting not working for me

### DISCUSSION THREAD

**Discussion by chrisp250** [rerdavies/pipedal#118] [Q&A] [answered]
Hi all,
Thank you for PiPedal. I've been using it with NAM and even though it has a few minor issues with NAM, I am having a blast with it. Specially now that the latest version fixed the issue showing the selected profile in NAM.

My query is about the sorting feature of the plugins list. When I go to add a plug in and ALL is selected, I can see all the plugins that PiPedal can find, however if I select any of the categories e.g. distortion, reverb, etc. nothing shows up. I went through all of the categories and they all show empty. What am I doing wrong?

Second question, when NAM builds and installs, it gets loaded in /usr/local/lib/lv2 by default. Even though the configuration for PiPedal shows as that being another folder to search for plugins, it doesn't pick it up. I had to do a symbolic link to /usr/lib/lv2 and fixed it. Has anyone else come across this?

Thank you

**Comment by rerdavies** [comment]
Are you sym-linking the entire `/usr/lib/lv2/` directory, or specifically symlinking the `/usr/lib/lv2/neural_amp_modeler.lv2/ `directory to `/usr/local/lib/lv2/neural_amp_modeler.lv2/ ` ? 

If the former, that would probably be the source of your problems with missing filter tags. 

Could you please verify for me that the following file is present:

    /usr/lib/lv2/core.lv2/lv2core.ttl


<looking at the path handling right now>.

**Reply by chrisp250** [reply]
Hi Robin,
Here is the sequence: first I did a make install for NAM. That copied the files to /usr/local/lib/lv2. I rebooted and PiPedal wouldn't see NAM. Then I symlinked /usr/local/lib/lv2/neural_amp_modeler.lv2 to /usr/lib/lv2, rebooted, and that fixed the issue. So just the NAM folder, not the whole /usr/local/lib/lv2.

No, I don't have a /usr/lib/lv2/core.lv2 folder. How could I fix that?

Thank you

**Comment by rerdavies** [comment]
At a bash prompt:
```
  ls  /usr/lib/lv2/neural_amp_modeler.lv2/*
```
What do you see?

**Reply by chrisp250** [reply]
This:
`/usr/lib/lv2/neural_amp_modeler.lv2/manifest.ttl           /usr/lib/lv2/neural_amp_modeler.lv2/neural_amp_modeler.ttl
/usr/lib/lv2/neural_amp_modeler.lv2/neural_amp_modeler.so`

This is now, having the symbolic link, as per:
`drwxr-xr-x 2 root root 4096 Jun 17 16:21 gxtubedelay.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_TubeDistortion.lv2
drwxr-xr-x 2 root root 4096 Jun 17 16:21 gxtubetremelo.lv2
drwxr-xr-x 2 root root 4096 Jun 17 16:21 gxtubevibrato.lv2
drwxr-xr-x 2 root root 4096 Jun 17 16:21 gxtuner.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_ultracab.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_uvox.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_valvecaster.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_vfm.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_vmk2d.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_voodoo.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_voxbass.lv2
drwxr-xr-x 2 root root 4096 Jun 17 15:11 gx_vstb.lv2
drwxr-xr-x 2 root root 4096 Jun 17 16:21 gx_w20.lv2
drwxr-xr-x 2 root root 4096 Jun 17 16:21 gx_zita_rev1.lv2
lrwxrwxrwx 1 root root   41 Jun 24 11:02 neural_amp_modeler.lv2 -> /usr/local/lib/lv2/neural_amp_modeler.lv2
drwxr-xr-x 6 root root 4096 Jun 25 09:52 ToobAmp.lv2
`

**Comment by rerdavies** [comment, marked as answer]
Could you please try the following at a bash prompt:
```
   sudo apt install lv2-dev
```
Restart the pipedal server:
```
    sudo pipedalconfig --restart
```

Close the pipedal web page, and then re-open it (to make sure all files and plugin information are reloaded)

Does that fix the plugin filters?

**Reply by chrisp250** [reply]
Perfect. That worked. It also gave me the icons for the modules that I didn't know were missing.

So it seems the dev package is required to get all the support files.

Thank you for your help.

**Reply by chrisp250** [reply]
Now I wonder if having had lv2-dev would have also solved the /usr/local/lib/lv2 issue...

**Comment by rerdavies** [comment]
Thanks for your problem report. The next version of pipedal will include lv2-dev as a dependency in the deb file, which will prevent this problem.

1 problem down. Now, back to the problem with loading plugins from /usr/local/lib/lv2....

I ran a test on my system. It does, in fact, find and load the NAM plugin when it's placed in `/usr/local/lv2`.

Theory #1. You do, in fact have to reload pipedal after isntalling a new plugin. LV2 plugins are scanned for once as pipedal loads. You can either reboot your Raspberry PI, OR you can issue the following command at a bash prompt to restart the service:
```
   sudo pipedalconfig -- restart
```

So. Let's try the following.... 
```
   sudo rm neural_amp_modeler.lv2     # remove the symlink.
   sudo pipedalconfig --restart
```
Close and re-open the PiPedal web page (not actually required this time, but let's be safe).

Does NAM load in pipedal?

ps. no. The missing lv2-dev package would not have caused problems with loading files from `/usr/local/lib/lv2`.

If that doesn't work, have you modified `/etc/pipedal/config.json`? 

Awaiting your response.

**Reply by chrisp250** [reply]
I deleted the symbolic link, rebooted, and everything still works. Seems I should have also closed the browser and not just rebooted the RPi when I installed NAM.

If folders like `core.lv2` are needed for the runtime operation of LV2, it seems like a poor choice by the distribution to include those in the dev package rather than the runtime package.

Thanks for your help!

**Reply by rerdavies** [reply]
Agreed. :-) I'm pinging the LV2 developers right now to see if there's a better package.  But those files are definitely required, and they are definitely shared by LV2 hosts.  And I can't find another package that contains them.

With respect to the requiring reloads and restarts to pick up newly installed plugins... I'm definitely unhappy with your user experience here.  PiPedal needs to do better. 

An F5 refresh of the browser page would also have worked fine. I think I can actually avoid that problem in the browser entirely, by reloading the plugin list after reconnecting. A fix for that will definitely show up in the next PiPedal release (sometime in the next couple of weeks).

Rescanning plugins on the fly is a bit more awkward. I need to give that some thought. But it seems fixable.

I have generated a bug report from this discussion. Expect to see improvements in the next release of Pipedal, which I expect to happen in the next couple of weeks. 

The rule of thumb we hold is that for every person who reports a bug, there a hundred who don't.  So your problem report is very much appreciated. Thank you again.

**Reply by chrisp250** [reply]
No problems at all. It's only a few minutes of effort compared to what you've put into it and we all get to enjoy it.

I posted on the NAM group on Facebook that NAM is working great on RPi4 and added a link to PiPedal. There seems to be a lot of interest in it. I thought I was going to have to stay confined to feather profiles, but no, I haven't had any issues with any profiles.

I also had a few issues with the last plugin in the chain. For some reason it stops processing the audio. If bypass it, audio works again. At the moment I have a reverb there, the way to fix it is to delete it and added again. Not sure if it's related to the actual plugin or it's just because it's the last effect in the chain. It's a bit hard to replicate. I could create another post for that if you want, but I'm afraid I'm not quite sure how to replicate the issue.

**Reply by rerdavies** [reply]
Dang! I've been seeing a HUGE spike in incoming traffic today. And github analytics don't say where they came from. That's probably it. Thank you.

I'll see if I can replicate the additional problem. If you get a repro, please post to a new thread, thanks.

If the question comes up, NAM still has a pending pull request from me that  addresses an issue of general concern to hosts.  The current 1.1.26 build of PiPedal is not affected by the issue.  There has, unfortunately, been a bit of thrashing over a fix for MODP that caused PiPedal some grief over the last week or so; but it's all settled out now, whether the last pull request goes in or not. You probably noticed that if you've been using fresh builds of NAM.

**Reply by chrisp250** [reply]
As you probably know, the LV2 NAM plug in is a fair bit down the development chain... iPlug2->NAM->NAM-LV2.
NAM would probably be classed beta, the NAM-LV2 would probably still be alpha, Mike just did his first release (0.1.0) only a few days ago, so I imagine things are still fairly green in this space.

You should join the NAM Facebook group if you have a Facebook account. I know the iPlug2 owner is there. Even though it's not really LV2 focussed, there are some interesting discussions there.

Great to see my post helped. Great timing as well with the update that fixed the atom:Path issue with NAM.

I'm also following the Zynthian project. They use mod there, and someone tried to get NAM working with poor results. I think it's because they are still on a 32bits platform. So as far as I know, PiPedal is the easiest way to get NAM running on RPi at the moment.
